projects/online_sequence_classification/clustering/clustering_interface.py

```diff
@@ -0,0 +1,282 @@
+  def stable_sequence(anomaly_score, stable_anomaly_score=0.2):
+    """
+    Determine whether a temporal sequence is stable.
+    
+    :param anomaly_score: (float) anomaly score of the temporal memory
+    :param stable_anomaly_score: (float) threshold to determine whether the 
+      anomaly score is stable.
+    :return: (bool) whether the sequence is stable 
+    """
+    if anomaly_score < stable_anomaly_score:
+      return True
+    else:
+      return False
+
+
+  def average_cluster_distance(self):
+    """
+    Average cluster distance between clusters.
+    
+    :return average_distance: (float) average distance between clusters. 
+    """
+    cluster_distances = []
+    cluster_ids = self.clusters.keys()
+    for i in cluster_ids:
+      for j in cluster_ids:
+        if i != j:
+          ci = self.clusters[i].center.value
+          cj = self.clusters[j].center.value
+          d = self.distance_func(ci, cj)
+          cluster_distances.append(d)
+
+    if len(cluster_distances) > 0:
+      return np.mean(cluster_distances)
+    else:
+      return 0.0
+
+
+  def add_or_merge_cluster(self, cluster, merge_threshold):
+    """
+    Add cluster to the existing clusters or merge it with the closest cluster.
+    
+    :param cluster: (Cluster) the cluster to assign
+    :param merge_threshold: (float) If the distance to the closest 
+      cluster is below the merge threshold, the cluster will be merged with 
+      the closest cluster. Otherwise, if the distance to the closest cluster 
+      is  above the merge threshold, then the cluster will be added to the 
+      existing clusters.
+    """
+    distance_to_closest, closest, = self.find_closest_cluster(cluster.center)
+    if closest and distance_to_closest < merge_threshold:
+      closest.merge(cluster)
+    else:
+      self.add_cluster(cluster)
+
+
+  def create_cluster(self, center=None):
+    """
+    Create a cluster.
+    
+    :param center: (Point) point to add.
+    """
+    cluster_id = len(self.clusters) + 1
+    cluster = Cluster(cluster_id, center)
+    return cluster
+
+
+  def add_cluster(self, cluster):
+    """
+    Add cluster to the existing clusters.
+    
+    :param cluster: (Cluster) cluster to add.
+    :raise: (ValueError) raise error if the cluster ID is already used. 
+    """
+    if cluster.id in self.clusters:
+      raise ValueError('Cluster ID %s already exists' % cluster.id)
+    self.clusters[cluster.id] = cluster
+
+
+  @abstractmethod
+  def find_closest_cluster(self, point):
+    """
+    Find the closest cluster to a point.
+    
+    :param point: (Point) The point of interest.
+    :return distance_to_closest: (float) distance between closest cluster 
+      center and point.
+    :return closest: (Cluster) closest cluster to point.
+    """
+    raise NotImplementedError()
+
+
+  def merge_closest_clusters(self):
+    """
+    Merge closest two clusters.
+    """
+    inter_cluster_dists = []
+    for c1 in self.clusters.values():
```

Comment:
I noticed that this loop goes through each pair of clusters twice. You could go through each pair once by:
```
numClusters = len(self.clusters)
for i in range(numCluster):
  for j in range(i+1, numClusters):
    c1 = self.clusters.values()[i]
    c2 = self.clusters.values()[j]
```


The same applies to the average_cluster_distance function